h2o-core/src/main/java/water/util/Log.java

```diff
@@ -97,105 +106,57 @@ public static void setLogLevel(String sLvl) {
-    org.apache.log4j.Logger l = LogManager.getLogger(water.api.RequestServer.class);
-    String s = String.format("  %-6s  %3d  %6d ms  %s", method, status, deltaMillis, uri);
-    l.info(s);
-  }
-
-  public static void info( String s, boolean stdout ) { if( _level >= INFO ) write0(INFO, stdout, new String[]{s}); }
+  public static void log  ( int level, Object... objs ) { write(level, objs); }
 
   // This call *throws* an unchecked exception and never returns (after logging).
   public static RuntimeException throwErr( Throwable e ) {
     err(e);                     // Log it
     throw e instanceof RuntimeException ? (RuntimeException)e : new RuntimeException(e); // Throw it
   }
 
-  private static void write( int lvl, Object objs[] ) {
-    boolean writeToStdout = (lvl <= _level);
-    write0(lvl, writeToStdout, objs);
+  private static void write(int lvl, Object[] objs) {
+    write0(lvl, objs);
   }
 
-  private static void write0( int lvl, boolean stdout, Object objs[] ) {
-    StringBuilder sb = new StringBuilder();
-    for( Object o : objs ) sb.append(o);
-    String res = sb.toString();
-    if( _bufferMessages ) { // Oops, need to buffer until we can do a proper header
-      INIT_MSGS.add(res);
-      return;
-    }
-    if( INIT_MSGS != null ) {   // Ahh, dump any initial buffering
-      String host = H2O.SELF_ADDRESS.getHostAddress();
-      _preHeader = fixedLength(host + ":" + H2O.API_PORT + " ", 22) + fixedLength(H2O.PID + " ", 6);
-      ArrayList<String> bufmsgs = INIT_MSGS;  INIT_MSGS = null;
-      if (bufmsgs != null) for( String s : bufmsgs ) write0(INFO, true, s);
+  private static void write0(int lvl, Object[] objs) {
+    StringBuilder msgBuff = new StringBuilder();
+    Throwable t = null;
+    for (int i = 0; i < objs.length - 1; i++) msgBuff.append(objs[i]);
+    if (objs.length > 0 && objs[objs.length - 1] instanceof Throwable) {
+      t = (Throwable) objs[objs.length-1];
+    } else {
+      msgBuff.append(objs[objs.length-1]);
     }
-    write0(lvl, stdout, res);
-  }
-
-  private static void write0( int lvl, boolean stdout, String s ) {
-    StringBuilder sb = new StringBuilder();
-    String hdr = header(lvl);   // Common header for all lines
-    write0(sb, hdr, s);
-
-    // stdout first - in case log4j dies failing to init or write something
-    if(stdout && !_quiet) System.out.println(sb);
-
-    // log something here
-    org.apache.log4j.Logger l4j = _logger != null ? _logger : createLog4j();
-    switch( lvl ) {
-    case FATAL:l4j.fatal(sb); break;
-    case ERRR: l4j.error(sb); break;
-    case WARN: l4j.warn (sb); break;
-    case INFO: l4j.info (sb); break;
-    case DEBUG:l4j.debug(sb); break;
-    case TRACE:l4j.trace(sb); break;
-    default:
-      l4j.error("Invalid log level requested");
-      l4j.error(s);
+    String msg = msgBuff.toString();
+    if (_bufferMessages) { // Oops, need to buffer until we can do a proper header
+      INIT_MSGS.add(new BufferedMsg(lvl, msg, t));
```

Comment:
does this preserve the information about the context? eg. the thread that logged the message?

Reply:
no, for such messages the context (time, thread) is lost, do you see that as an issue?

Reply:
IMHO not a big issue, typically it will be logged by the main thread anyway

it could help with debugging what called an exit during the start-up process... but not a big issue I think